### PR TITLE
tool: separating ceph-dedup-tool into tool and daemon  

### PR DESCRIPTION
**Pull request by myoungwon**
https://github.com/ceph/ceph/pull/53992

 - PR 1 : separating dedup tool and daemon (commit 1-3)

 - PR 2 : renaming and refactoring global variable

 - PR 3 : rework deduplication procedure in a async. manner
 
 - PR 4 : handling dedup-daemon's parameter and config + tests and subconscious bug fixes (commit 4-18)

 - PR 5 : Depending on the progress, integration to cephadm? or unittest framework (commit 19~)

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/tools/ceph_dedup/ceph_dedup_daemon.cc [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #include "common.h"
 
 ceph::mutex glock = ceph::make_mutex("glock");
+class SampleDedupWorkerThread;
+std::list<SampleDedupWorkerThread> threads;
+bool all_stop = false;
```
Comment (athanatos) [inline src/tools/ceph_dedup/ceph_dedup_daemon.cc:6]: Before moving onto your next work item, please submit a PR refactoring these globals into an automated variable declared in main().

File: src/tools/ceph_dedup/ceph_dedup_daemon.cc [hunk 2]
```diff
@@ -0,0 +1,817 @@
...
+  if (g_conf()->daemonize) {
+    global_init_postfork_finish(g_ceph_context);
+    forker.daemonize();
+  }
+
+  init_async_signal_handler();
+  register_async_signal_handler_oneshot(SIGINT, handle_signal);
+  register_async_signal_handler_oneshot(SIGTERM, handle_signal);
+
+  int ret = make_crawling_daemon(opts);
```
Comment (athanatos) [inline src/tools/ceph_dedup/ceph_dedup_daemon.cc:810]: Before your next work item, submit a PR renaming make_crawling_daemon to run_crawling_daemon.

File: src/tools/ceph_dedup/ceph_dedup_daemon.cc [hunk 3]
```diff
@@ -0,0 +1,817 @@
...
+  if (opts.count("run-once")) {
+    run_once = true;
+  }
+
+  cout << "SampleRatio : " << sampling_ratio << std::endl
+    << "Chunk Dedup Threshold : " << chunk_dedup_threshold << std::endl
+    << "Chunk Size : " << chunk_size << std::endl
+    << std::endl;
+
+  while (!all_stop) {
```
Comment (athanatos) [inline src/tools/ceph_dedup/ceph_dedup_daemon.cc:672]: Reading all_stop here is undefined behavior as glock is not held.

File: src/tools/ceph_dedup/ceph_dedup_daemon.cc [hunk 4]
```diff
@@ -0,0 +1,817 @@
...
+  bool stop = false;
+  ceph::mutex m_lock = ceph::make_mutex("SampleDedupWorkerThread");
+};
+
+void SampleDedupWorkerThread::crawl()
+{
+  cout << "new iteration" << std::endl;
+
+  ObjectCursor current_object = begin;
+  while (!stop && current_object < end) {
```
Comment (athanatos) [inline src/tools/ceph_dedup/ceph_dedup_daemon.cc:334]: Reading stop here is undefined behavior as m_lock is not held.  In a larger sense -- why bother checking stop here and not all_stop?  Why add m_lock/stop in addition to g_lock/all_stop?

**Review by athanatos — CHANGES_REQUESTED** [review]
I stopped reviewing for now after the above comments.  Before re-requesting review please do the following:
1. Add a commit adding comment to every member or variable accessed in more than one thread ```// Accessed in the main thread and in the worker threads under <lock_name>```
2. Ensure that for every one of those variables, the same lock is held for all accesses -- both reads and writes.
3. Ensure that there are as few of the above as possible

Also, do not complicate the locking structure by adding more locks or by adding shared-exclusive locks for now -- let's focus on getting the simple version correct first.

File: src/tools/ceph_dedup/ceph_dedup_daemon.cc (hunk 2 above)
Comment (myoungwon) [reply src/tools/ceph_dedup/ceph_dedup_daemon.cc:824]: OK.

**Comment by myoungwon** [comment]
@athanatos Can you take a look at https://github.com/ceph/ceph/pull/56067/commits/c34a055edf6d8f4c7210f02907c535bcdc72d0c9 and https://github.com/ceph/ceph/pull/56067/commits/8129d826d47ef6b11f95f719b48a986a8fb9f685 to see if I have addressed your comments?

**Comment by myoungwon** [comment]
> Can we drop "tool/ceph_dedup: rework deduplication procedure in the daemon" and introduce it in its own PR?

@athanatos OK. Done.

**Comment by athanatos** [comment]
jenkins test signed

**Review by athanatos — APPROVED** [review]
Ok, lgtm once it goes through testing.  Once this is scheduled for a run and the run completes, please confirm that the run includes your modified test and that the test passes.  Ping me again once that happens and I'll take care of doing the merge.

This review identified 2-3 small changes that should be their own PRs, let's do those next (one at a time, please).

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues65126

**Pull request merged**